src/parser.rs

```diff
@@ -37,17 +39,25 @@ impl fmt::Display for EncodedChunk {
     }
 }
 
-#[derive(Debug, PartialEq, Eq)]
 /// Things that can go wrong when restoring a chunked file
+#[derive(Debug, PartialEq, Eq, thiserror::Error)]
 pub enum RestoreError {
     /// Not enough chunks for the expected total
+    #[error(
+        "Missing some chunks! Expected to see {expected_total} chunks, but missing chunks: \
+         {missing_chunk_ids:?}"
+    )]
     MissingChunk {
         /// How many we thought we'd find
```

Comment:
Looks amazing. Am I understanding this correctly that this is just syntactic sugar based on macros?

Also what's with derive under vs above the docstring? Which one should it be normally? I was wondering about this before

Reply:
correct! this is a procedural macro that writes pretty much the implementation you already had, plus a couple of extras (it also implements `std::error::Error` for your struct).

it's capable of a few other tricks too, such as implementing `From` other error types for you.

as for the order of docstrings above or below derive.. it doesn't really matter. docstring before attributes is *preferred* but not enforced. there's an open issue for rustfmt to add a configuration option for this, which would be off by default